src/rust/mangastream/template/src/helper.rs

```diff
@@ -359,3 +373,105 @@ pub fn get_lang_code() -> String {
 	}
 	String::new()
 }
+
+static mut CACHED_MANGA_URL_TO_POSTID_MAPPING: Option<HashMap<String, String>> = None;
+static mut CACHED_MAPPING_AT: f64 = 0.0;
+
+// This requests the "all manga" listing page in text mode and parses out
+// the postid and url for each manga, and caches it in a hashmap to prevent
+// having to request the page again.
+//
+// The all manga listing page is the only reliable way to get the postids for
+// each manga, without making a request to each and every manga page when
+// browsing (*cough* paperback *cough*)
+//
+/// Generate a hashmap of manga url to postid mappings
+fn generate_manga_url_to_postid_mapping(url: &str, pathname: &str) -> HashMap<String, String> {
+	unsafe {
+		// if the mapping was generated less than 10 minutes ago, use the cached mapping
+		if current_date() - CACHED_MAPPING_AT < 600.0 {
+			if let Some(mapping) = &CACHED_MANGA_URL_TO_POSTID_MAPPING {
+				return mapping.clone();
+			}
+		}
+	}
+
+	let all_manga_listing_url = format!("{}/{}/list-mode", url, pathname);
+
+	let html = Request::get(all_manga_listing_url)
+		.html()
+		.expect("Failed to get html");
+
+	let mut mapping = HashMap::new();
+
+	for node in html.select(".soralist .series").array() {
+		let manga = node.as_node().expect("Failed to get node");
+
+		let url = manga.attr("href").read();
+		let post_id = manga.attr("rel").read();
+
+		mapping.insert(url, post_id);
+	}
+
+	unsafe {
+		CACHED_MANGA_URL_TO_POSTID_MAPPING = Some(mapping.clone());
+		CACHED_MAPPING_AT = current_date();
+	}
+
+	mapping
+}
+
+/// Search the `MANGA_URL_TO_POSTID_MAPPING` for the postid from a manga url
+pub fn get_postid_from_manga_url(url: String, base_url: &str, pathname: &str) -> String {
+	let manga_url_to_postid_mapping = generate_manga_url_to_postid_mapping(base_url, pathname);
+
+	String::from(
+		manga_url_to_postid_mapping
+			.get(&url)
+			.expect("Failed to get post id from url"),
```

Comment:
same here

Reply:
Had to have an `AidokuError` and none of them seemed appropriate, so I just went with `AidokuErrorKind::Unimplemented`